go/parquet/schema/column.go

```diff
@@ -0,0 +1,125 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schema
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/apache/arrow/go/parquet"
+	format "github.com/apache/arrow/go/parquet/internal/gen-go/parquet"
+)
+
+// Column encapsulates the information necessary to interpret primitive
+// column data in the context of a particular schema. We have to examine
+// the node structure of a column's path to the root in the schema tree
+// to be able to reassemble the nested structure from the repetition and
+// definition levels.
+type Column struct {
+	pnode     *PrimitiveNode
+	maxDefLvl int16
+	maxRepLvl int16
+}
+
+func (c *Column) hasSpacedValues() bool {
+	if c.maxRepLvl > 0 {
+		// repeated + flat case
+		return c.pnode.RepetitionType() != parquet.Repetitions.Required
+	}
+
+	// non-repeated+nested case
+	// find if a node forces nulls in the lowest level along the hierarchy
+	var n Node = c.pnode
+	for n != nil {
+		if n.RepetitionType() == parquet.Repetitions.Optional {
```

Comment:
Does this mirror current C++ code.  I think once you hit a repeated field you can also stop.

Reply:
ah NM there can be no repeated fields here.

Reply:
I'll take a look, I wrote this back in january so it's possible that the C++ has been improved since then and i missed it.

Reply:
not sure if you saw comment above.  I think this is right since there is no repeated fields here.  I don't think this affects correctness but I think the check above isn't quite correct here or in C++.  really we need to walk the tree in both cases. Until either the top or a repeated field is hit

Reply:
although really, if this isn't related to arrow code, I don't think we need this method here, reading spaced values isn't something lower level parquet reading should really be concerned with

Reply:
Looking some more at the C++ code, it appears HasSpacedValues is only used in ReadBatchSpaced which is deprecated precisely because a lot of this logic is broken, and outside of Arrow (which no longer uses this code path) doesn't serve much value.

Reply:
so this looks like an artifact from earlier versions of my code, I didn't notice that I don't call it anymore and even my go-staticcheck is pointing out that this is no longer called at all, even in my finished version of the repo. So i'll remove this function entirely.